### PR TITLE
feat(evs/volume): support prepaid resource creation

### LINKED ISSUE #1844: [EVS] The volume does not support periodic payment
Hi team,

The v2 API supports the periodic payment feature. After the API version is updated from v3 to v2, this feature can be programmed.
https://support.huaweicloud.com/api-evs/evs_04_2003.html

Best regards

### PR DESCRIPTION
**Pull request by Lance52259**
**What this PR does / why we need it**:
For EVS volumes, the API support prepaid resource creation.
new params:
- charging_mode
- period_unit
- period
- auto_renew

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #1844 

**Special notes for your reviewer**:

**Release note**:

```release-note
1. add prepaid of charging mode support for EVS volume.
```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST='./huaweicloud/services/acceptance/evs' TESTARGS='-run=TestAccEvs
Volume_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/evs -v -run=TestAccEvsVolume_basic -timeout 360m -parallel 4
=== RUN   TestAccEvsVolume_basic
=== PAUSE TestAccEvsVolume_basic
=== CONT  TestAccEvsVolume_basic
--- PASS: TestAccEvsVolume_basic (133.27s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/evs     133.337s
```

```
make testacc TEST='./huaweicloud/services/acceptance/evs' TESTARGS='-run=TestAccEvsVolume_prePaid'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/evs -v -run=TestAccEvsVolume_prePaid -timeout 360m -parallel 4
=== RUN   TestAccEvsVolume_prePaid
=== PAUSE TestAccEvsVolume_prePaid
=== CONT  TestAccEvsVolume_prePaid
--- PASS: TestAccEvsVolume_prePaid (152.02s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/evs 152.090s
```

### CODE REVIEW

File: huaweicloud/services/evs/resource_huaweicloud_evs_volume.go [hunk 1]
```diff
@@ -208,22 +231,35 @@ func resourceEvsVolumeCreate(ctx context.Context, d *schema.ResourceData, meta i
...
-	if err != nil {
-		return fmtp.DiagErrorf("Error waiting for the creation of EVS volume (%s) to complete: %s", d.Id(), err)
+	// If charging mode is PrePaid, wait for the order to be completed.
+	if strings.EqualFold(d.Get("charging_mode").(string), "prePaid") {
+		err = common.WaitOrderComplete(ctx, d, config, job.OrderID)
+		if err != nil {
+			return fmtp.DiagErrorf("[DEBUG] Error the order is not completed while creating EVS volume (%s): %#v",
+				d.Id(), err)
+		}
+	} else {
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/evs/resource_huaweicloud_evs_volume.go]: Is it confirmed that the disk is available when the order is complete?

File: huaweicloud/services/evs/resource_huaweicloud_evs_volume.go [hunk 2]
```diff
@@ -402,16 +446,23 @@ func resourceEvsVolumeDelete(ctx context.Context, d *schema.ResourceData, meta i
...
-	// If this is true, just move on. It'll eventually delete.
-	if v.Status != "deleting" {
-		if err := cloudvolumes.Delete(evsV2Client, d.Id(), deleteOpts).ExtractErr(); err != nil {
-			return common.CheckDeletedDiag(d, err, "EVS volume")
+	if d.Get("charging_mode").(string) == "prePaid" {
+		err = common.UnsubscribePrePaidResource(d, config, []string{d.Id()})
+		if err != nil {
+			return fmtp.DiagErrorf("error unsubscribing HuaweiCloud EVS volume : %s", err)
+		}
+	} else {
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/evs/resource_huaweicloud_evs_volume.go:459]: should we try to delete the disk after unsubcribe it?

File: huaweicloud/services/evs/resource_huaweicloud_evs_volume.go [hunk 3]
```diff
@@ -402,16 +446,23 @@ func resourceEvsVolumeDelete(ctx context.Context, d *schema.ResourceData, meta i
...
+		// The snapshots associated with the disk are deleted together with the EVS disk if cascade value is true
+		deleteOpts := cloudvolumes.DeleteOpts{
+			Cascade: d.Get("cascade").(bool),
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/evs/resource_huaweicloud_evs_volume.go:462]: What about the snapshots associated with the prepaid disk?

**Comment by niuzhenguo** [comment]
shoud also consider when expanding the size of prepaid resource, there will be a new order for it.

File: huaweicloud/services/evs/resource_huaweicloud_evs_volume.go (hunk 2 above)
Comment (Lance52259) [reply huaweicloud/services/evs/resource_huaweicloud_evs_volume.go:459]: Prepaid disks will be completely removed after the order is completed.

File: huaweicloud/services/evs/resource_huaweicloud_evs_volume.go (hunk 3 above)
Comment (Lance52259) [reply huaweicloud/services/evs/resource_huaweicloud_evs_volume.go:462]: The `cascade` parameter is only useful for postpaid disks, and I will add a note to the document.

File: huaweicloud/services/evs/resource_huaweicloud_evs_volume.go [hunk 4]
```diff
@@ -184,17 +203,21 @@ func buildEvsVolumeCreateOpts(d *schema.ResourceData, config *config.Config) clo
 		volumeOpts.Metadata = m
 	}
```
Comment (ShiChangkuo) [inline huaweicloud/services/evs/resource_huaweicloud_evs_volume.go]: it seems that the metadata at least has one key.

File: huaweicloud/services/evs/resource_huaweicloud_evs_volume.go [hunk 5]
```diff
@@ -149,6 +153,19 @@ func ResourceEvsVolume() *schema.Resource {
...
 }
 
+func buildBssParamParams(d *schema.ResourceData) *cloudvolumes.BssParam {
+	bp := cloudvolumes.BssParam{
+		ChargingMode: d.Get("charging_mode").(string),
+	}
+	if strings.EqualFold(bp.ChargingMode, "prePaid") {
```
Comment (ShiChangkuo) [inline huaweicloud/services/evs/resource_huaweicloud_evs_volume.go]: the caller has checked this condition, can we skip it?

File: huaweicloud/services/evs/resource_huaweicloud_evs_volume.go [hunk 6]
```diff
@@ -208,22 +231,35 @@ func resourceEvsVolumeCreate(ctx context.Context, d *schema.ResourceData, meta i
...
+	if len(job.VolumeIDs) < 1 {
+		return fmtp.DiagErrorf("The volume ID was not included in the response to the request to create the volume.")
 	}
+	d.SetId(job.VolumeIDs[0])
 
-	resp, err := stateConf.WaitForStateContext(ctx)
-	if err != nil {
-		return fmtp.DiagErrorf("Error waiting for the creation of EVS volume (%s) to complete: %s", d.Id(), err)
+	// If charging mode is PrePaid, wait for the order to be completed.
+	if strings.EqualFold(d.Get("charging_mode").(string), "prePaid") {
```
Comment (ShiChangkuo) [inline huaweicloud/services/evs/resource_huaweicloud_evs_volume.go]: maybe `job.OrderID != ""` is better

File: huaweicloud/services/evs/resource_huaweicloud_evs_volume.go (hunk 5 above)
Comment (Lance52259) [reply huaweicloud/services/evs/resource_huaweicloud_evs_volume.go]: yeah, that is a useless check.

File: huaweicloud/services/evs/resource_huaweicloud_evs_volume.go (hunk 1 above)
Comment (Lance52259) [reply huaweicloud/services/evs/resource_huaweicloud_evs_volume.go]: done

File: huaweicloud/services/evs/resource_huaweicloud_evs_volume.go (hunk 6 above)
Comment (Lance52259) [reply huaweicloud/services/evs/resource_huaweicloud_evs_volume.go]: seems great.

File: huaweicloud/services/evs/resource_huaweicloud_evs_volume.go (hunk 4 above)
Comment (Lance52259) [reply huaweicloud/services/evs/resource_huaweicloud_evs_volume.go]: OFC

**Comment by Lance52259** [comment]
The useless client (BlockStorageV1Client) and releated function have been removed from `config.go` and `endpoint_test.go`.

**Pull request merged**